### DISCUSSION TITLE
Avatar Cascade Delete

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#9722] [Archive]
When a user has an avatar i can't delete it. First i have to delete the file of the image.
Is there a config with which this happens automaticly ?

I have written a small hook for this, but i hope there is a setting for this.

**Comment by rijkvanzanten** [comment]
There's no way to do that automatically at this point. We can't rely on the set null foreign key constraint here, as MS SQL doesn't support multiple columns that relate to the same foreign table with constraints (cc @jaycammarano).

The plan here is to have Directus show a dedicated dialog to warn about the foreign key constraint, and have the user chose how to proceed. Until that's the case, you can either delete the file manually first, or rely on a custom hook to do it automatically 👍🏻

**Reply by NilsBaumgartner1994** [reply]
Mhh okay thanks. I think a small section of "reciepes" collections for hooks like these would be very good.

**Reply by rijkvanzanten** [reply]
Totally! Lemme mark this as a good addition for the docs 👍🏻

**Reply by NilsBaumgartner1994** [reply]
A further idea would be, that from these "reciepes" community addons will be created via verce. Like "and I want that addon also"

**Reply by NilsBaumgartner1994** [reply]
> There's no way to do that automatically at this point. We can't rely on the set null foreign key constraint here, as MS SQL doesn't support multiple columns that relate to the same foreign table with constraints (cc @jaycammarano).
> 
> The plan here is to have Directus show a dedicated dialog to warn about the foreign key constraint, and have the user chose how to proceed. Until that's the case, you can either delete the file manually first, or rely on a custom hook to do it automatically 👍🏻

@rijkvanzanten How to write a filter hook with nested relations?
City -> has many Buildings -> Has Image

When i delete a city, all buildings will automatically get deleted, but they won't fire the filter(buildings.items.delte) hook?
I would have to write a filter hook for city and manually check every possible combination where images may be used in sub-collections?

**Comment by NilsBaumgartner1994** [comment]
```
const AvatarHelper = require("../AvatarHelper");

/**
 * A small hook to delete the avatar image before user deletion
 */
module.exports = function registerHook({ filter, action, init, schedule }, { services, exceptions, database, getSchema, env, logger }) {
  // when a user has an avatar, it cant be deleted, so we need to delete the file ourself with a hook
  filter('users.delete', async (payload, input, {database, schema, accountability}) => {
    let usersIds = payload; //get the user ids
    for(let userId of usersIds){ // for all users which get deleted
      await AvatarHelper.deleteAvatarOfUser(services, database, schema, accountability, exceptions, userId) //delete avatar file
    }

    return payload;
  });

};
```





```
class AvatarHelper{

  /**
   * Deletes the avatar file for a userId
   * @param userId the userId
   * @returns {Promise<void>}
   */
  static async deleteAvatarOfUser(services, database, schema, accountability, exceptions, userId){

    const filesService = AvatarHelper.getAdminFileServiceInstance(schema, accountability, services);
    const { InvalidPayloadException } = exceptions;
    if(!userId){
      throw new InvalidPayloadException("deleteAvatarOfUser: No userId provided: ");
    }

    const existingUser = await database("directus_users").where({ id: userId }).first(); //get user
    if(!existingUser){ //handle no user found error
      throw new InvalidPayloadException("deleteAvatarOfUser: No user found with id: "+userId);
    }

    const avatar_filename = existingUser.avatar; //get filename of avatar
    if(!!avatar_filename){ //if has image
      await filesService.deleteOne(avatar_filename); //delete file
    }
  }

  /**
   * get a fileService with admin permission
   * @returns {*}
   */
  static getAdminFileServiceInstance(schema, accountability, services){
    const { FilesService } = services;
    let adminAccountAbility = AccountHelper.getAdminAccountability(accountability);
    const filesService = new FilesService({ schema, accountability: adminAccountAbility });
    return filesService;
  }

}
```


```
/**
 * Helper for Account things
 */
class AccountHelper{

  /**
   * Set the accountablility to admin
   * @param accountability the given accountability
   * @returns {any} a copy of the accountability with admin permission
   */
  static getAdminAccountability(accountability){
    let adminAccountAbility = JSON.parse(JSON.stringify(accountability)); //make a copy !
    adminAccountAbility.admin = true; //usefull if we realy want to upload something as admin
    return adminAccountAbility;
  }

}

module.exports = AccountHelper;
```

**Reply by NilsBaumgartner1994** [reply]
Put them in Hooks
